### PR TITLE
Add the support of IBM Power and ARM architectures for wkhtmltopdf on Odoo 16

### PR DESCRIPTION
**Pull request by ludovic-gasc**
Hi,

This pull request takes care of the CPU architecture to be able to install the right architecture version for the wkhtmltopdf package.

We have tested with Podman on RHEL 9 + IBM Power and RedHat OpenShift on IBM Power.

I have also added one line for arm64 support, but we have not tested it.

Thanks for your feedback and the merge.

### CODE REVIEW

**Comment by samip5** [comment]
Did you also test that PDF generation works eg invoice preview/print? I had the problem with Debian based image that I created for multi-arch, but it went away with using fedora 36 and the RPM from nightly package repo.

File: 16.0/Dockerfile [hunk 1]
```diff
@@ -6,35 +6,44 @@ SHELL ["/bin/bash", "-xo", "pipefail", "-c"]
...
+    python3-pyldap \
+    python3-qrcode \
+    python3-renderpm \
+    python3-setuptools \
+    python3-slugify \
+    python3-vobject \
+    python3-watchdog \
+    python3-xlrd \
+    python3-xlwt \
+    xz-utils && \
```
Comment (d-fence) [inline 16.0/Dockerfile:38]: I know that the indentation is historically broken, but can you keep it that way to avoid unnecessary diff.

File: 16.0/Dockerfile [hunk 2]
```diff
@@ -6,35 +6,44 @@ SHELL ["/bin/bash", "-xo", "pipefail", "-c"]
...
+    WKHTMLTOPDF_ARCH=${TARGETARCH} && \
+    case ${TARGETARCH} in \
+    "amd64") WKHTMLTOPDF_SHA=9df8dd7b1e99782f1cfa19aca665969bbd9cc159  ;; \
+    "arm64")  WKHTMLTOPDF_SHA=58c84db46b11ba0e14abb77a32324b1c257f1f22  ;; \
+    "ppc64le") WKHTMLTOPDF_ARCH=ppc64el && WKHTMLTOPDF_SHA=7ed8f6dcedf5345a3dd4eeb58dc89704d862f9cd  ;; \
+    esac \
+    && curl -o wkhtmltox.deb -sSL https://github.com/wkhtmltopdf/packaging/releases/download/0.12.6.1-3/wkhtmltox_0.12.6.1-3.bullseye_${WKHTMLTOPDF_ARCH}.deb \
```
Comment (d-fence) [inline 16.0/Dockerfile:45]: It would be better to keep the [known good version](https://github.com/odoo/odoo/wiki/Wkhtmltopdf#version-status) of wkhtmltopdf for the amd64 platform. And use the 0.12.6.1 for arm64 and ppc64le as  0.12.5 is not packaged for those platforms.

**Review by d-fence — CHANGES_REQUESTED** [review]
Thanks for this nice pull request.
Just a few changes and it should be good.

**Comment by d-fence** [comment]
Tested with a cross build linux/ppc64le and it worked. Same for linux/arm64.
We are testing wkhtmltopdf 0.12.6.1-3 and at a first glance, it seems to be ok. But, in order to avoid breaking existing  deployements, I think that it would be better to keep 0.12.5 for the amd64 platform.

**Comment by ludovic-gasc** [comment]
Hi @d-fence ,

Thanks for the tests, I have seen your comments.

About your comment using 0.12.5 instead of 0.12.6 for amd64, I fully understand the risk of regression, nevertheless, I permit you to point to the changelog: https://github.com/wkhtmltopdf/wkhtmltopdf/releases/tag/0.12.6
And the diff between 0.12.5 and 0.12.6.1:
https://github.com/wkhtmltopdf/wkhtmltopdf/compare/0.12.5...0.12.6

The changes are very minimal and mainly to support ppc64le and ARM.
Moreover, I'm not a big fan of using different versions depending on the architecture, we might have another behavior depending on the architecture, not the best for a multi-architecture application.
Last but not least, the release date of 0.12.5 is 2018, no packages for the new stable version of Debian, 12 - bookworm, but it seems we have working bookworm packages for 0.12.6: https://github.com/wkhtmltopdf/packaging/releases/tag/0.12.6.1-3

The day you change the Debian image in the Dockerfile for the latest Debian stable, you must also use the latest version of wkhtmltopdf.

I might have another suggestion:
Maybe for the next stable release of Odoo, 17, you might change the version recommended for wkhtmltopdf?
Moreover, we might put the official support of the two new architectures only for Odoo 17?

Thanks for your feedback, and in case you prefer to keep 0.12.5 for amd64, then I will adapt my pull request.

**Comment by d-fence** [comment]
Hello,

I agree that the changes are minimal and the basic tests that we made did not reveal any issue.
So let's go with  0.12.6.

I just ask you to remove the indentation fix to avoid useless diff.

Thanks

File: 16.0/Dockerfile (hunk 1 above)
Comment (ludovic-gasc) [reply 16.0/Dockerfile]: Fixed with this commit: https://github.com/odoo/docker/commit/f576ee2922745ac3b8000f549929896f111345cf

File: 16.0/Dockerfile (hunk 2 above)
Comment (ludovic-gasc) [reply 16.0/Dockerfile:45]: As discussed here: https://github.com/odoo/docker/pull/466#issuecomment-1741484925
We will keep for now 0.12.6.1 for amd64.
Always possible to add an exception in the future in case a regression is detected.

**Comment by ludovic-gasc** [comment]
Hi @d-fence ,

I have made the changes.
FYI, we have planned to do more tests next week on IBM Power, just to be sure it's OK.

Thanks for your comments

**Comment by d-fence** [comment]
OK I will wait for the tests

**Comment by ludovic-gasc** [comment]
Hi @d-fence ,

We have done more tests and we confirm the PDF generation works pretty well on IBM Power and RedHat OpenShift.
A few PDF examples generated on IBM Power are in the attachment.

[INV-2023-00004 (1).pdf](https://github.com/odoo/docker/files/12832829/INV-2023-00004.1.pdf)
[Production Order - WH_MO_00001.pdf](https://github.com/odoo/docker/files/12832834/Production.Order.-.WH_MO_00001.pdf)
[Picking Operations - Wood Corner - WH_OUT_00007.pdf](https://github.com/odoo/docker/files/12832836/Picking.Operations.-.Wood.Corner.-.WH_OUT_00007.pdf)
[Delivery Slip - Wood Corner - WH_OUT_00007.pdf](https://github.com/odoo/docker/files/12832837/Delivery.Slip.-.Wood.Corner.-.WH_OUT_00007.pdf)

Could you merge this pull request ?

Thanks.

**Comment by ludovic-gasc** [comment]
> Did you also test that PDF generation works eg invoice preview/print? I had the problem with Debian based image that I created for multi-arch, but it went away with using fedora 36 and the RPM from nightly package repo.

Hi @samip5 ,

Yes, we have tested invoice preview/print, it was the main goal of this pull request :-)
See my previous comment with a few PDF examples.

I have used the Debian package from the creator of this tool instead of the Debian repository: https://github.com/ludovic-gasc/docker/blob/ppc64le/16.0/Dockerfile#L45C56-L45C56

If you could test this Dockerfile on ARM, it would be awesome, I have no ARM motherboard to be able to test.

Kind regards

**Comment by ludovic-gasc** [comment]
Hi @ashuio ,

Let me test from my side if I have the same issue, I will test this week.

Thanks for the potential bug report.

**Comment by ludovic-gasc** [comment]
Hi @ashuio , I have just tested on Ubuntu 22.04.1 and Docker 24.0.5 with snap, as suggested by the installer of Ubuntu, and it works pretty well, cf. screenshot:
image

Maybe your installation has an issue ?
How you have installed Docker on this Ubuntu ?
Could you try again, maybe on another system?

Thanks

**Comment by ludovic-gasc** [comment]
hi @d-fence , do you still see something to change before merging this pull request ?

If not, it would be wonderful if you could merge this PR :-)

Thanks, and have a nice week!

**Comment by ludovic-gasc** [comment]
hi @d-fence ,

Good catch for the empty variable.

We are waiting the official publication on DockerHub to do extra tests on IBM Power to be sure everything will be OK.

Thank you a lot for your help

**Comment by d-fence** [comment]
> Good catch for the empty variable.

I had to change it anyway because the docker hub doesn't use TARGETARCH.
So I added a fixing commit to use dpkg --print-architecture as [recommended](https://github.com/docker-library/official-images#multiple-architectures)

**Comment by hoangtrann** [comment]
I tested this PR on a Mac running M1 Pro and it works like a charm. To generate the PDF, I set the `report.url` to `http://0.0.0.0:8069` and `wkhtmltopdf` works without any problems.

**Comment by ludovic-gasc** [comment]
hi @d-fence , thank you a lot for your awesome help:
Now that Docker Inc has merged your pull request, I have tested the official new Odoo container image on the top of RedHat OpenShift and IBM Power, it works as expected.

I confirm that the ARM and IBM Power support is now visible on the Docker Hub:
https://hub.docker.com/search?q=odoo&image_filter=official

I am now this pull request.

Have a nice week
